tests/30rooms/60version_upgrade.pl

```diff
@@ -373,6 +373,81 @@ sub upgrade_room_synced {
       });
    };
 
+test "/upgrade preserves the power level of the upgrading user in old and new rooms",
+   requires => [
+      local_user_and_room_fixtures(),
+      local_user_fixture(),
+      qw( can_upgrade_room_version can_change_power_levels ),
+   ],
+
+   do => sub {
+      my ( $creator, $room_id, $upgrader ) = @_;
+
+      my ( $pl_content, $new_room_id );
+
+      # Note that this test assumes that moderators by default are allowed to upgrade rooms
+
+      matrix_join_room_synced(
+         $upgrader, $room_id
+      )->then( sub {
+         # Make the joined user a moderator
+         matrix_change_room_power_levels(
+            $creator, $room_id, sub {
+               ( $pl_content ) = @_;
+               $pl_content->{users}->{$upgrader->user_id} = JSON::number(50);
+               log_if_fail "PL content in old room", $pl_content;
+            }
+         )
+      })->then( sub {
+         matrix_sync( $upgrader );
+      })->then( sub {
+         upgrade_room_synced(
+            $upgrader, $room_id,
+            expected_event_counts => { 'm.room.power_levels' => 1 },
+            new_version => $TEST_NEW_VERSION,
+         );
+      })->then( sub {
+         ( $new_room_id, ) = @_;
+
+         matrix_sync_again( $upgrader );
+      })->then( sub {
+         my ( $sync_body ) = @_;
+
+         log_if_fail "sync body", $sync_body;
+
+         # Two power level events will be sent in the new room. The first is to make the
+         # upgrader user (previously a moderator) an Administrator (which can only be done
+         # when creating the room). This is such that they could send the initial state
+         # events. The second power level event is to downgrade the upgrader user from a
+         # Administrator to a Moderator again to keep a consistent state with the old room
+
+         # Check the contents of the latest power level event
+         my $new_room = $sync_body->{rooms}{join}{$new_room_id};
+         my $new_room_pl_event = first {
+            $_->{type} eq 'm.room.power_levels'
+         } reverse @{ $new_room->{timeline}->{events} };
+
+         log_if_fail "PL event in new room", $new_room_pl_event;
+
+         # Check that the power levels in the new room match the original PLs
+         assert_deeply_eq(
+            $new_room_pl_event->{content},
+            $pl_content,
+            "power levels in replacement room",
+         );
+
+         # Check that the power levels in the old room have not changed
+         # by making sure there is not a power levels event for the old room
+         # in the sync response
+         my $old_room = $sync_body->{rooms}{join}{$room_id};
```

Comment:
have you checked this actually failed under the old code?

My suspicion is that it will not, since the bug is that we are modifying the existing PL event, rather than creating a new one. Maybe try fetching the state with `/r0/room/{roomId}/state/m.room.power_levels` or sth?

Reply:
It does fail with:

```
 | Got JSON::number(100), expected JSON::number(50) at {users}{@anon-20200114_100001-7:localhost:8800} for power levels in replacement room at tests/30rooms/60version_upgrade.pl line 435.
```

But perhaps fetching state is a more reliable method anyways?

Reply:
Switched. Test still fails on old code, passes on new.